src/go/sysfail.go

```diff
@@ -0,0 +1,186 @@
+)
+
+type ThreadDiscoveryConfig struct {
+	PollIntervalUsec uint32
+}
+
+type SyscallOutcome struct {
+	Syscall int
+	Outcome Outcome
+}
+
+type Plan struct {
+	Strategy        ThreadDiscoveryStrategy
+	Config          ThreadDiscoveryConfig
+	Ctx             unsafe.Pointer
+	Selector        func(ctx unsafe.Pointer, tid C.sysfail_tid_t) int
+	SyscallOutcomes []*SyscallOutcome
+}
+
+type Session struct {
+	session *C.sysfail_session_t
+	plan    *C.sysfail_plan_t
+}
+
+// NewPlan creates a new plan for failure injection.
+func NewPlan(strategy ThreadDiscoveryStrategy, pollIntervalUsec uint32, syscallOutcomes []*SyscallOutcome) *Plan {
+	return &Plan{
+		Strategy: strategy,
+		Config: ThreadDiscoveryConfig{
+			PollIntervalUsec: pollIntervalUsec,
+		},
+		SyscallOutcomes: syscallOutcomes,
+		Selector:        nil, // Use appropriate selector function
+		Ctx:             nil, // User data
+	}
+}
+
+// StartSession starts a new failure injection session.
+func StartSession(plan *Plan) *Session {
+	var cOutcomes *C.sysfail_syscall_outcome_t
+	if len(plan.SyscallOutcomes) > 0 {
+		cOutcomes = (*C.sysfail_syscall_outcome_t)(C.malloc(C.size_t(len(plan.SyscallOutcomes) * int(unsafe.Sizeof(C.sysfail_syscall_outcome_t{})))))
+		var prev *C.sysfail_syscall_outcome_t = nil
+		for i, outcome := range plan.SyscallOutcomes {
+			curr := (*C.sysfail_syscall_outcome_t)(unsafe.Pointer(uintptr(unsafe.Pointer(cOutcomes)) + uintptr(i)*unsafe.Sizeof(*cOutcomes)))
+			curr.syscall = C.int(outcome.Syscall)
+			curr.outcome.fail.p = C.double(outcome.Outcome.Fail.P)
+			curr.outcome.fail.after_bias = C.double(outcome.Outcome.Fail.AfterBias)
+			curr.outcome.delay.p = C.double(outcome.Outcome.Delay.P)
+			curr.outcome.delay.after_bias = C.double(outcome.Outcome.Delay.AfterBias)
+			curr.outcome.max_delay_usec = C.uint(outcome.Outcome.MaxDelayUsec)
+			curr.outcome.ctx = outcome.Outcome.Ctx
+			// curr.outcome.eligible = // still need to convert the function pointer if used
+			curr.outcome.num_errors = C.uint(outcome.Outcome.NumErrors)
+			if outcome.Outcome.NumErrors > 0 {
+				ewSize := int(unsafe.Sizeof(C.sysfail_error_wt_t{}))
+				var cErrors *C.sysfail_error_wt_t
+				cErrors = (*C.sysfail_error_wt_t)(C.malloc(C.size_t(outcome.Outcome.NumErrors) * C.size_t(ewSize)))
+				for j, ew := range outcome.Outcome.ErrorWeights {
+					ewPtr := (*C.sysfail_error_wt_t)(unsafe.Pointer(uintptr(unsafe.Pointer(cErrors)) + uintptr(j)*unsafe.Sizeof(*cErrors)))
+					ewPtr.nerror = C.int(ew.Nerror)
+					ewPtr.weight = C.double(ew.Weight)
+				}
+				C.set_error_weights(&curr.outcome, cErrors)
+			}
+			if prev != nil {
+				prev.next = curr
+			}
+			prev = curr
+		}
+	}
+
+	cPlan := &C.sysfail_plan_t{
+		strategy:         C.sysfail_thread_discovery_strategy_t(plan.Strategy),
+		ctx:              plan.Ctx,
+		selector:         nil, // Use appropriate selector function
+		syscall_outcomes: cOutcomes,
+	}
+
+	C.set_poll_itvl(&cPlan.config, C.uint(plan.Config.PollIntervalUsec))
```

Comment:
Make this conditional on polling strategy?

When we support netlink discovery this may become a bug (iirc config is a union).